sparsepp gets flagged with so many `cast-user-defined` warnings on GCC 15 that I can't get all of my CI build log for my project

With gcc 15 as shipped in Ubuntu 26.04, there are new compiler warnings introduced. The build of my project using sparsepp now produces such a volume of warnings that my CI log reaches the log limit size I have of 4 megabytes, and I'm not able to see the end.

I need to reduce the amount of warnings like this one that GCC 15 raises when I do operations on types from sparsepp.

The warning is `cast-user-defined`. This is supposed to "Warn when a cast to reference type does not involve a user-defined conversion that the programmer might expect to be called.", according to https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html.

It is upset that, when sparsepp casts from `value_type` (an `std::pair` with a const first element) to `mutable_reference` (a reference to a pair with a non-const first element), it isn't going to use what appears to be the pair copy constructor. I'm not sure why it thinks the pair copy constructor is a user-defined conversion operator.

The GCC documentation does not explain what they think code is supposed to be doing instead; it might be that it's supposed to use C++-style `static_cast<>` constructs?

Here's an example of the type of warning I need sparsepp to stop raising in such large volumes on GCC 15:

```
#22 383.8 /vg/include/sparsepp/spp.h: In instantiation of 'void spp::sparsegroup<T, Alloc>::_init_val(mutable_value_type*, reference) [with T = std::pair<const handlegraph::net_handle_t, long unsigned int>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; mutable_value_type = std::pair<handlegraph::net_handle_t, long unsigned int>; reference = std::pair<const handlegraph::net_handle_t, long unsigned int>&]':
#22 383.8 /vg/include/sparsepp/spp.h:1411:22:   required from 'void spp::sparsegroup<T, Alloc>::_set_aux(allocator_type&, size_type, Val&, realloc_not_ok_type) [with Val = std::pair<const handlegraph::net_handle_t, long unsigned int>; T = std::pair<const handlegraph::net_handle_t, long unsigned int>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; allocator_type = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = unsigned char; realloc_not_ok_type = spp::integral_constant<bool, false>]'
#22 383.8  1411 |             _init_val((mutable_pointer)&_group[num_items], val);
#22 383.8       |             ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:1439:21:   required from 'void spp::sparsegroup<T, Alloc>::_set(allocator_type&, size_type, size_type, Val&) [with Val = std::pair<const handlegraph::net_handle_t, long unsigned int>; T = std::pair<const handlegraph::net_handle_t, long unsigned int>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; allocator_type = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = unsigned char]'
#22 383.8  1439 |             _set_aux(alloc, offset, val, check_alloc_type());
#22 383.8       |             ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:1457:13:   required from 'spp::sparsegroup<T, Alloc>::value_type* spp::sparsegroup<T, Alloc>::set(allocator_type&, size_type, Val&) [with Val = std::pair<const handlegraph::net_handle_t, long unsigned int>; T = std::pair<const handlegraph::net_handle_t, long unsigned int>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; pointer = std::pair<const handlegraph::net_handle_t, long unsigned int>*; allocator_type = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = unsigned char]'
#22 383.8  1457 |         _set(alloc, i, offset, val);            // may change _group pointer
#22 383.8       |         ~~~~^~~~~~~~~~~~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:2255:27:   required from 'void spp::sparsetable<T, Alloc>::move(size_type, reference) [with T = std::pair<const handlegraph::net_handle_t, long unsigned int>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = long unsigned int; reference = std::pair<const handlegraph::net_handle_t, long unsigned int>&]'
#22 383.8  2255 |         which_group(i).set(_alloc, pos_in_group(i), val);
#22 383.8       |         ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:2834:23:   required from 'void spp::sparse_hashtable<Value, Key, HashFcn, ExtractKey, SetKey, EqualKey, Alloc>::_move_from(MoveDontCopyT, spp::sparse_hashtable<Value, Key, HashFcn, ExtractKey, SetKey, EqualKey, Alloc>&, size_type) [with Value = std::pair<const handlegraph::net_handle_t, long unsigned int>; Key = handlegraph::net_handle_t; HashFcn = vg::wang_hash<handlegraph::net_handle_t>; ExtractKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SelectKey; SetKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SetKey; EqualKey = std::equal_to<handlegraph::net_handle_t>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = long unsigned int]'
#22 383.8  2834 |             table.move(bucknum, *it);    // moves the value to here
#22 383.8       |             ~~~~~~~~~~^~~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:2952:9:   required from 'spp::sparse_hashtable<Value, Key, HashFcn, ExtractKey, SetKey, EqualKey, Alloc>::sparse_hashtable(MoveDontCopyT, spp::sparse_hashtable<Value, Key, HashFcn, ExtractKey, SetKey, EqualKey, Alloc>&, size_type) [with Value = std::pair<const handlegraph::net_handle_t, long unsigned int>; Key = handlegraph::net_handle_t; HashFcn = vg::wang_hash<handlegraph::net_handle_t>; ExtractKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SelectKey; SetKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SetKey; EqualKey = std::equal_to<handlegraph::net_handle_t>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = long unsigned int]'
#22 383.8  2952 |         _move_from(mover, ht, min_buckets_wanted);
#22 383.8       |         ^~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:2684:30:   required from 'bool spp::sparse_hashtable<Value, Key, HashFcn, ExtractKey, SetKey, EqualKey, Alloc>::_maybe_shrink() [with Value = std::pair<const handlegraph::net_handle_t, long unsigned int>; Key = handlegraph::net_handle_t; HashFcn = vg::wang_hash<handlegraph::net_handle_t>; ExtractKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SelectKey; SetKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SetKey; EqualKey = std::equal_to<handlegraph::net_handle_t>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >]'
#22 383.8  2684 |             sparse_hashtable tmp(MoveDontCopy, *this, sz);
#22 383.8       |                              ^~~
#22 383.8 /vg/include/sparsepp/spp.h:2850:13:   required from 'void spp::sparse_hashtable<Value, Key, HashFcn, ExtractKey, SetKey, EqualKey, Alloc>::resize(size_type) [with Value = std::pair<const handlegraph::net_handle_t, long unsigned int>; Key = handlegraph::net_handle_t; HashFcn = vg::wang_hash<handlegraph::net_handle_t>; ExtractKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SelectKey; SetKey = spp::sparse_hash_map<handlegraph::net_handle_t, long unsigned int, vg::wang_hash<handlegraph::net_handle_t>, std::equal_to<handlegraph::net_handle_t>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::SetKey; EqualKey = std::equal_to<handlegraph::net_handle_t>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = long unsigned int]'
#22 383.8  2850 |             _maybe_shrink();
#22 383.8       |             ^~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:3803:51:   required from 'void spp::sparse_hash_map<Key, T, HashFcn, EqualKey, Alloc>::resize(size_type) [with Key = handlegraph::net_handle_t; T = long unsigned int; HashFcn = vg::wang_hash<handlegraph::net_handle_t>; EqualKey = std::equal_to<handlegraph::net_handle_t>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = long unsigned int]'
#22 383.8  3803 |     void resize(size_type cnt)        { rep.resize(cnt); }
#22 383.8       |                                         ~~~~~~~~~~^~~~~
#22 383.8 /vg/include/sparsepp/spp.h:3805:41:   required from 'void spp::sparse_hash_map<Key, T, HashFcn, EqualKey, Alloc>::reserve(size_type) [with Key = handlegraph::net_handle_t; T = long unsigned int; HashFcn = vg::wang_hash<handlegraph::net_handle_t>; EqualKey = std::equal_to<handlegraph::net_handle_t>; Alloc = spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> >; size_type = long unsigned int]'
#22 383.8  3805 |     void reserve(size_type cnt)       { resize(cnt); } // c++11
#22 383.8       |                                         ^~~~~~
#22 383.8 src/subcommand/../snarl_seed_clusterer.hpp:450:57:   required from here
#22 383.8   450 |                 net_handle_to_node_problem_index.reserve(5*seed_count);
#22 383.8       |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
#22 383.8 /vg/include/sparsepp/spp.h:1343:40: warning: casting 'spp::sparsegroup<std::pair<const handlegraph::net_handle_t, long unsigned int>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::value_type' {aka 'std::pair<const handlegraph::net_handle_t, long unsigned int>'} to 'spp::sparsegroup<std::pair<const handlegraph::net_handle_t, long unsigned int>, spp::libc_allocator<std::pair<const handlegraph::net_handle_t, long unsigned int> > >::mutable_reference' {aka 'std::pair<handlegraph::net_handle_t, long unsigned int>&'} does not use 'constexpr std::pair<_T1, _T2>::pair(const std::pair<_U1, _U2>&) [with _U1 = const handlegraph::net_handle_t; _U2 = long unsigned int; typename std::enable_if<(std::_PCC<((! std::is_same<_T1, _U1>::value) || (! std::is_same<_T2, _U2>::value)), _T1, _T2>::_ConstructiblePair<_U1, _U2>() && std::_PCC<((! std::is_same<_T1, _U1>::value) || (! std::is_same<_T2, _U2>::value)), _T1, _T2>::_ImplicitlyConvertiblePair<_U1, _U2>()), bool>::type <anonymous> = true; _T1 = handlegraph::net_handle_t; _T2 = long unsigned int]' [-Wcast-user-defined]
#22 383.8  1343 |         ::new (p) value_type(std::move((mutable_reference)val));
#22 383.8       |                                        ^~~~~~~~~~~~~~~~~~~~~~
```

--- a/sparsepp/spp.h
+++ b/sparsepp/spp.h
@@ -1340,7 +1340,7 @@ private:
     void _init_val(mutable_value_type *p, reference val)
     {
 #if !defined(SPP_NO_CXX11_RVALUE_REFERENCES)
-        ::new (p) value_type(std::move((mutable_reference)val));
+        ::new (p) value_type(std::move(*((mutable_pointer)(&val))));
 #else
         ::new (p) value_type((mutable_reference)val);
 #endif
@@ -1356,7 +1356,7 @@ private:
     void _set_val(value_type *p, reference val)
     {
 #if !defined(SPP_NO_CXX11_RVALUE_REFERENCES)
-        *(mutable_pointer)p = std::move((mutable_reference)val);
+       *(mutable_pointer)p = std::move(*((mutable_pointer)(&val)));
 #else
         using std::swap;
         swap(*(mutable_pointer)p, *(mutable_pointer)&val);

--- a/tests/spp_test.cc
+++ b/tests/spp_test.cc
@@ -2738,7 +2738,11 @@ struct Memmove
 public:
     Memmove(): i(0) {}
     explicit Memmove(int ival): i(ival) {}
-    Memmove(const Memmove& that) { this->i = that.i; num_copies++; }
+    Memmove(const Memmove& that) {
+       this->i = that.i;
+       num_copies++;
+    }
+    Memmove& operator=(const Memmove&) = default;
     int i;
     static int num_copies;
 };
@@ -2750,6 +2754,7 @@ public:
     NoMemmove(): i(0) {}
     explicit NoMemmove(int ival): i(ival) {}
     NoMemmove(const NoMemmove& that) { this->i = that.i; num_copies++; }
+    NoMemmove& operator=(const NoMemmove&) = default;
     int i;
     static int num_copies;
 };

